woocommerce/woocommerce-ios
# Order Creation: Support negative values in shipping and fees

Closes: https://github.com/woocommerce/woocommerce-ios/issues/6352.

## Description

This PR updates "Shipping Details" and "Fee Details" to support negative values.

## Changes

- Updates `PriceFieldFormatter` to support negative numbers
- Updates `FeeLineDetailsViewModel` with negative amount and percentage support
- Updates `ShippingLineDetailsViewModel` with negative amount support

### Known limitations

iOS doesn't provide a number pad with negative input option. The current solution is to use `.numbersAndPunctuation` keyboard (see screenshots below). We can improve it by adding "+/-" button in `inputAccessoryView` above the numpad. It'll require a change in our nested UIKit component (`BindableTextfield`).

## Testing

1. Build and run the app in debug mode (to ensure the feature flag is enabled).
2. Make sure Order Creation is enabled under Settings > Experimental Features.
3. Go to the Orders tab and tap the + button. Tap "Create order" in the bottom sheet.
4. Add a product (to enable percentage fees).
5. Add a negative shipping amount.
6. Check the shipping line in the Payment section.
7. Check that negative amount is prefilled when shipping is edited.
8. Add a negative fee.
9. Check the fee line in the Payment section.
10. Check that negative amount is prefilled when fee is edited.
11. Update fee to a percentage type, save negative amount.

## Screenshots

negative shipping|negative fee|negative payment section
--|--|--
![Simulator Screen Shot - 1](https://user-images.githubusercontent.com/3132438/157100433-cc11f17e-3281-4454-8209-3138ce398972.png)|![Simulator Screen Shot - 2](https://user-images.githubusercontent.com/3132438/157100569-8b5e3511-49e0-4927-a6c6-5efc7319b4c0.png)|![Simulator Screen Shot - 3](https://user-images.githubusercontent.com/3132438/157100578-0b95ff4e-921a-420a-8391-3b793a5784cc.png)

---
